architecture/crds.md

```diff
@@ -0,0 +1,324 @@
+
+After looking at several ways of making it easy for chart developers to do this, we alighted on a fairly easy (though inelegant) solution: Helm, as it stands today, provides a separate location in a chart to add CRDs. These will be uploaded _before the chart's templates are even rendered_ (thus avoiding verification issues with the discovery API). If the CRD exists or if the permissions do not allow it, the chart will fail before the application is deployed.
+
+Of course, users are unhappy with this for a host of reasons. They want CRDs templated (without understanding the race conditions). They want stronger version controls. They don't like having a separate directory for CRDs. We sympathize... but we currently have not devised a better solution for the aforementioned problems.
+
+> NOTE: There is no requirement that CRDs can only be placed in the `crd/` directory. They can be put along side other resources in the `templates/` directory. This was an intentional design choice to preserve backward compatibility.
+
+## Deleting CRDs
+
+We'll delay talking about upgrades for just a moment, and skip to the easiest one: Deleting CRDs.
+
+Helm currently does not delete CRDs. The reasoning behind this decision is the trivial confluence of two aspects of CRDs: global resources and cascading deletes.
+
+The next two subsections explain these.
+
+### Shared Global  Resources
+
+Earlier, we looked at how Kubernetes CRDs are _shared global resources_. That is, only one resource describes the CRD and all of its versions, and that one resource is shared among all namespaces on a cluster.
+
+If application A and application B both rely on the same CRD, allowing application A to delete the CRD would be bad news for application B. While it's fairly harmless to allow the creation of a CRD by the first thing that wants it, it is bad to allow any dependent application to delete that CRD when that application cannot confirm that in so doing it will not cripple other applications that use the same CRD.
+
+However, there is actually no way to tell whether or not anything uses a CRD. That is an internal implementation detail of the controllers, which is entirely opaque to cluster operators, let alone to Helm. The Kubernetes API server _could_ remedy this situation, but does not. (Example: controllers could be required to indicate to the Kubernetes API server which CRDs they rely upon.)
+
+Even if Helm could determine whether a CRD was consumed by multiple Kubernetes applications, it could not tell which usages were required (application _depends_ on the CRD), optional (application can use the CRD if present), or incidental (application merely collects data). Again, this is because Kubernetes itself does not provide facilities for this (though it could).
```

Comment:
It sounds like you have an idea how this could be done, I'd love to hear it and see if I can find some bandwidth to solve it.

Reply:
At the very least, this is what we need: A way to query which things are listening for a particular CRD, and what their in-cluster ID is (pod UUID/name/namespace is sufficient, since we can trace that back through replica sets and deployments/statefulsets/etc). External listeners should be reported, too. But understandably they would not have a unique ID that ties them to a Pod.

There are obviously a variety of ways Kubernetes itself could capture that information. I don't have any preferences. And I think even if this were an optional feature add to the existing event streams, developers would use it because it makes CRD management easier.